### PR TITLE
feat: XCALL Executor

### PR DESCRIPTION
**Pull request by debendraoli**
> Closes #109 #134 #133 #99

### CODE REVIEW

File: relayer/chains/wasm/provider.go [hunk 1]
```diff
@@ -0,0 +1,746 @@
...
+
+func (p *Provider) getRawContractMessage(message *relayTypes.Message) (wasmTypes.RawContractMessage, error) {
+	switch message.EventType {
+	case events.EmitMessage:
+		rcvMsg := types.NewExecRecvMsg(message)
+		return json.Marshal(rcvMsg)
+	case events.CallMessage:
+		execMsg := types.NewExecExecMsg(message)
+		return json.Marshal(execMsg)
+	case events.RevertMessage:
```
Comment (sherpalden) [inline relayer/chains/wasm/provider.go:631]: Are we monitoring this event: `RevertMessage`. I don't see its implementation in the event listener.

File: relayer/chains/wasm/provider.go [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+
+func (p *Provider) ShouldReceiveMessage(ctx context.Context, message *relayTypes.Message) (bool, error) {
+	return true, nil
+}
+
+func (p *Provider) ShouldSendMessage(ctx context.Context, message *relayTypes.Message) (bool, error) {
+	return true, nil
+}
+
+func (p *Provider) GenerateMessages(ctx context.Context, messageKey *relayTypes.MessageKeyWithMessageHeight) ([]*relayTypes.Message, error) {
```
Comment (sherpalden) [inline relayer/chains/wasm/provider.go:440]: Finality processor is not needed in cosmos based chains.

Comment (debendraoli) [reply relayer/chains/wasm/provider.go:439]: Needed to ensure packet are delivered irrespective of finality.

File: relayer/chains/wasm/provider.go [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+}
+
+func (p *Provider) Route(ctx context.Context, message *relayTypes.Message, callback relayTypes.TxResponseFunc) error {
+	p.logger.Info("starting to route message", zap.Any("message", message))
+	res, err := p.call(ctx, message)
+	if err != nil {
+		return err
+	}
+	go p.waitForTxResult(ctx, message.MessageKey(), res.TxHash, callback)
+	seq := p.wallet.GetSequence() + 1
```
[the remark this answers is not in the document]
Comment (debendraoli) [reply relayer/chains/wasm/provider.go:154]: Its Deterministically  safe as there will be not much messages to process in single block.

**Review by sherpalden — DISMISSED** [review]

File: relayer/chains/wasm/provider.go [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+func (p *Provider) RevertMessage(ctx context.Context, sn *big.Int) error {
+	msg := relayTypes.Message{
+		Sn:        sn.Uint64(),
+		EventType: events.RevertMessage,
+	}
+	_, err := p.call(ctx, &msg)
+	return err
+}
+
+func (p *Provider) SetAdmin(ctx context.Context, address string) error {
```
[the remark this answers is not in the document]
Comment (debendraoli) [reply relayer/chains/wasm/provider.go]: Of course it is updated from cmd, see `use` command.

File: relayer/chains/wasm/types/types.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+type TxSearchParam struct {
+	StartHeight, EndHeight uint64
+	Events                 []types.Event
+	Prove                  bool
+	Page                   *int
+	PerPage                *int
+	OrderBy                string
+}
+
+func (param *TxSearchParam) BuildQuery() string {
```
[the remark this answers is not in the document]
Comment (debendraoli) [reply relayer/chains/wasm/types/types.go:21]: Unable to reproduce. 

Still its not a problem, at the end of day all the packets are received and delivered.

File: relayer/chains/icon/provider.go [hunk 6]
```diff
@@ -8,108 +8,134 @@ import (
...
-	KeyStore        string `json:"keystore" yaml:"keystore"`
-	Password        string `json:"password" yaml:"password"`
-	StartHeight     uint64 `json:"start-height" yaml:"start-height"` // would be of highest priority
-	ContractAddress string `json:"contract-address" yaml:"contract-address"`
-	NetworkID       uint   `json:"network-id" yaml:"network-id"`
-	NID             string `json:"nid" yaml:"nid"`
+type Config struct {
+	ChainName     string                         `json:"-" yaml:"-"`
+	RPCUrl        string                         `json:"rpc-url" yaml:"rpc-url"`
+	Address       string                         `json:"address" yaml:"address"`
```
[the remark this answers is not in the document]
Comment (debendraoli) [reply relayer/chains/icon/provider.go:20]: Address is generic of course its what it meant for, relayer address.

File: relayer/chains/icon/provider.go [hunk 7]
```diff
@@ -8,108 +8,134 @@ import (
...
+	StartHeight   uint64                         `json:"start-height" yaml:"start-height"` // would be of highest priority
+	BlockInterval string                         `json:"block-interval" yaml:"block-interval"`
+	Contracts     relayerTypes.ContractConfigMap `json:"contracts" yaml:"contracts"`
+	NetworkID     uint                           `json:"network-id" yaml:"network-id"`
+	FinalityBlock uint64                         `json:"finality-block" yaml:"finality-block"`
```
[the remark this answers is not in the document]
Comment (debendraoli) [reply relayer/chains/icon/provider.go:25]: Finality does not only for finality check also for packets verification.

File: relayer/chains/icon/provider.go [hunk 8]
```diff
@@ -118,3 +144,24 @@ func (icp *IconProvider) RevertMessage(ctx context.Context, sn *big.Int) error {
 	}
 	return nil
 }
+
+// SetAdmin sets the admin address of the bridge contract
```
[the remark this answers is not in the document]
Comment (debendraoli) [reply relayer/chains/icon/provider.go:148]: Again, setAdmin is called from cmd only, its done there.

**Review by sherpalden — APPROVED** [review]

**Pull request merged**